File: src/test/java/TestQueue.java
import java.time.Duration;
import java.util.concurrent.atomic.AtomicInteger;
import com.alexmherrmann.util.hedgehog.queue.Asyncifier;
import com.alexmherrmann.util.hedgehog.queue.QueueImpl;
import kotlinx.coroutines.AwaitKt;
import org.awaitility.Awaitility;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class TestQueue {
  @Test
  public void testAsync() {
    var queue = new QueueImpl<Integer>();

    var goal = new AtomicInteger(0);
    var async = Asyncifier.asyncifyReceiver(queue);

    var sum = new AtomicInteger(0);
    async.jSubscribe(sum::addAndGet);

    for (int i = 0; i < 50; i++) {
      queue.send(i);
      goal.addAndGet(i);
    }


    Awaitility
      .await()
      .atMost(Duration.ofSeconds(3))
      .until(() -> sum.get() == goal.get());
  }

}

File: src/test/kotlin/TestQueue.kt
package com.alexmherrmann.util.hedgehog.queue;

import com.alexmherrmann.util.hedgehog.queue.Asyncifier
import com.alexmherrmann.util.hedgehog.queue.Dealable
import com.alexmherrmann.util.hedgehog.queue.Squeese
import io.kotest.assertions.timing.eventually
import io.kotest.matchers.shouldBe
import kotlinx.coroutines.runBlocking
import java.util.*
import java.util.concurrent.ArrayBlockingQueue
import kotlin.test.Test
import kotlin.test.asserter
import kotlin.time.Duration.Companion.seconds

class QueueImpl<T> : Squeese<T> {
	private class DealableImpl<T>(
		private val obj: T,
	) : Dealable<T> {
		override fun markAsDealtWith() = Unit
		override fun getObj(): T = obj
	}
	val que = ArrayBlockingQueue<T>(200)
	override fun send(obj: T) {
		que.put(obj)
	}

	override fun receive(): Optional<Dealable<T>> = que.poll().toOptional().map { DealableImpl(it) }
}

class TestQueue {
	@Test
	fun simpleTest() {
		val queue = QueueImpl<Int>()
		queue.send(1)
		queue.send(2)
		queue.send(3)


		asserter.assertEquals("got 1", 1, queue.receive().get())
		asserter.assertEquals("got 2", 2, queue.receive().get())
		asserter.assertEquals("got 3", 3, queue.receive().get())
		asserter.assertEquals("got nothing", Optional.empty<Int>(), queue.receive())
		asserter.assertEquals("got nothing", Optional.empty<Int>(), queue.receive())

	}


	@Test
	fun asyncTest() {
		val queue = QueueImpl<Int>()
		val asyncQueue = Asyncifier.asyncifyReceiver(queue)
		var sum = 0
		asyncQueue.subscribe { sum += it }

		val asyncSender = Asyncifier.asyncifySender(queue)


		val range = 0..<50
		range.forEach {
			asyncSender
				.sendAsync(it)
				.thenAccept {
					it.isSuccessful() shouldBe true
				}
		}

		range.sum() shouldBe 1225


		runBlocking {
			eventually(1.seconds) {
				sum shouldBe 1225
			}
		}


	}
}

fun <T> T?.toOptional(): Optional<T> =
	Optional.ofNullable(this) as Optional<T>